distributed/s3fs.py

```diff
@@ -0,0 +1,424 @@
+# -*- coding: utf-8 -*-
+import io
+import logging
+import re
+
+import boto3
+from botocore.exceptions import ClientError
+from botocore.client import Config
+
+from dask.base import tokenize
+from .utils import read_block
+
+logger = logging.getLogger(__name__)
+
+logging.getLogger('boto3').setLevel(logging.WARNING)
+logging.getLogger('botocore').setLevel(logging.WARNING)
+
+def split_path(path):
+    """
+    Normalise S3 path string into bucket and key.
+
+    Parameters
+    ----------
+    path : string
+        Input path, like `s3://mybucket/path/to/file`
+
+    Examples
+    --------
+    >>> split_path("s3://mybucket/path/to/file")
+    ("mybucket", "path/to/file")
+    """
+    if path.startswith('s3://'):
+        path = path[5:]
+    if '/' not in path:
+        return path, ""
+    else:
+        return path.split('/', 1)
+
+
+class S3FileSystem(object):
+    """
+    Access S3 data as if it were a file system.
+    """
+    _conn = {}
+    connect_timeout=5
+    read_timeout=15
+
+    def __init__(self, anon=True, key=None, secret=None, **kwargs):
+        """
+        Create connection object to S3
+
+        Will use configured key/secret (typically in ~/.aws, see the
+        boto3 documentation) unless specified
+
+        Parameters
+        ----------
+        anon : bool (True)
+            whether to use anonymous connection (public buckets only)
+        key : string (None)
+            if not anonymouns, use this key, if specified
+        secret : string (None)
+            if not anonymous, use this password, if specified
+        kwargs : other parameters for boto3 session
+        """
+        self.anon = anon
+        self.key = key
+        self.secret = secret
+        self.kwargs = kwargs
+        self.connect(anon, key, secret, kwargs)
+        self.dirs = {}
+        self.s3 = self.connect(anon, key, secret, kwargs)
+
+    def connect(self, anon, key, secret, kwargs):
+        tok = tokenize(anon, key, secret, kwargs)
+        if tok not in self._conn:
+            logger.debug("Open S3 connection.  Anonymous: %s",
+                         self.anon)
+            if self.anon:
+                from botocore import UNSIGNED
+                conf = Config(connect_timeout=self.connect_timeout,
+                              read_timeout=self.read_timeout,
+                              signature_version=UNSIGNED)
+                s3 = boto3.Session().client('s3', config=conf)
```

Comment:
The non-anon version below allows for extra kwargs. We never test any (I don't know what they could be).